### PR TITLE
Isolate task update methods

### PR DESCRIPTION
**Pull request by mcornaton**
First step of https://github.com/iExecBlockchainComputing/iexec-core/issues/434: isolation of task update methods

### CODE REVIEW

File: src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java [hunk 1]
```diff
@@ -132,7 +136,7 @@ public ReplicateSupplyService(ReplicatesService replicatesService,
             // no need to ge further if the consensus is already reached on-chain
             // the task should be updated since the consensus is reached but it is still in RUNNING status
             if (taskService.isConsensusReached(task)) {
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java:138]: Why keeping isConsensusReached in TaskService? Should a now pure TaskService depends on ReplicateService?

File: src/main/java/com/iexec/core/task/TaskService.java [hunk 2]
```diff
@@ -53,26 +45,16 @@
     private final TaskUpdateRequestManager taskUpdateRequestManager;
     private final IexecHubService iexecHubService;
     private final ReplicatesService replicatesService;
-    private final ApplicationEventPublisher applicationEventPublisher;
-    private final WorkerService workerService;
-    private final BlockchainAdapterService blockchainAdapterService;
 
     public TaskService(
             TaskRepository taskRepository,
             TaskUpdateRequestManager taskUpdateRequestManager,
```
Comment (jeremyjams) [inline src/main/java/com/iexec/core/task/TaskService.java:51]: Do we have taskUpdateRequestManager usages here?

File: src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java (hunk 1 above)
Comment (mcornaton) [reply src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java:138]: Not sure about this one. I was thinking about putting only transitional methods into `TaskUpdateManager`. `isConsensusReached` is still a Task-related methods but does not make a task transition from one state to another so I guessed it was still belonging to `TaskService`.
Do you think it should be in another service?

File: src/main/java/com/iexec/core/task/TaskService.java (hunk 2 above)
Comment (mcornaton) [reply src/main/java/com/iexec/core/task/TaskService.java:51]: We don't, thanks for the catch!

File: src/main/java/com/iexec/core/task/TaskUpdateManager.java [hunk 3]
```diff
@@ -0,0 +1,612 @@
...
+     */
+    @Override
+    public void onTaskUpdateRequest(String chainTaskId) {
+        log.info("Received task update request [chainTaskId:{}]", chainTaskId);
+        this.updateTaskRunnable(chainTaskId);
+    }
+
+    public CompletableFuture<Boolean> publishUpdateTaskRequest(String chainTaskId) {
+        return taskUpdateRequestManager.publishRequest(chainTaskId);
+    }
```
Comment (zguesmi) [inline src/main/java/com/iexec/core/task/TaskUpdateManager.java:88]: Why do we use the same service to publish and consume at the same time? I think `publishUpdateTaskRequest()` is redundant, we should use `TaskUpdateRequestManager` directly to publish an update request.

File: src/main/java/com/iexec/core/task/TaskUpdateManager.java [hunk 4]
```diff
@@ -0,0 +1,612 @@
...
+                             BlockchainAdapterService blockchainAdapterService) {
+        this.taskService = taskService;
+        this.taskRepository = taskRepository;
+        this.taskUpdateRequestManager = taskUpdateRequestManager;
+        this.iexecHubService = iexecHubService;
+        this.replicatesService = replicatesService;
+        this.applicationEventPublisher = applicationEventPublisher;
+        this.workerService = workerService;
+        this.blockchainAdapterService = blockchainAdapterService;
+        this.taskUpdateRequestManager.setRequestConsumer(this);
```
Comment (zguesmi) [inline src/main/java/com/iexec/core/task/TaskUpdateManager.java:74]: With the comment below we wouldn't need to do this.

File: src/main/java/com/iexec/core/task/TaskUpdateManager.java (hunk 3 above)
Comment (mcornaton) [reply src/main/java/com/iexec/core/task/TaskUpdateManager.java:88]: That was the previous behavior in `TaskService` that I only copied. See: https://github.com/iExecBlockchainComputing/iexec-core/blob/6375324c701cf36b858c7483d716ba34107bc2ee/src/main/java/com/iexec/core/task/TaskService.java#L209

I don't exactly know why it has been designed this way. @jeremyjams, any clue on that?
If there's no reason to that publish/consume behavior to happen there, we could certainly remove it.

Comment (zguesmi) [reply src/main/java/com/iexec/core/task/TaskUpdateManager.java:88]: Ok, we can keep it that way in this PR, and eventually change it after.

**Comment by zguesmi** [comment]
Sorry, it was not on purpose.

**Review by zguesmi — APPROVED** [review]
"Let the purge begin!"

File: src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java (hunk 1 above)
Comment (jeremyjams) [reply src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java:138]: 2 points:
A - Why having this pre-transition check in a method (in a dedicated service) and still having other pre-transition checks with booleans without method? (non consistency)
B - This method doesn't need an access to the TaskRepo while other methods do (static-like)

Comment (mcornaton) [reply src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java:138]: A - It's also used in `ReplicateSupplyService` which is different from other pre-transition checks booleans.
B - You're right.

I see your point and it's probably better for it to be in `TaskUpdateManager`. I'll move it. Thanks!

Comment (jeremyjams) [reply src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java]: OK for ReplicateSupplyService thanks.
We might see if a separate checker would make sense

File: src/main/java/com/iexec/core/task/TaskUpdateManager.java (hunk 3 above)
Comment (jeremyjams) [reply src/main/java/com/iexec/core/task/TaskUpdateManager.java:88]: We can either 
(1) Merge TaskUpdateManager & TaskUpdateRequestManager (but testing is trickier)
or
(2) Use TaskUpdateRequestManager as a botleneck/filter (feature) between any client service and the TaskUpdateManager. Doing so, no one but the TaskUpdateRequestManager will be able to talk to the TaskUpdateManager  (to ensures TaskUpdateManager is fed properly)

**Pull request merged**